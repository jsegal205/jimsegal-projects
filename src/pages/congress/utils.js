export const chamberTitles = (chamber) => {
  switch (chamber) {
    case "house":
      return "Representatives";
    case "senate":
      return "Senators";
    default:
      return "Members";
  }
};

export const individualChamberTitle = (chamber) => {
  const pluralTitle = chamberTitles(chamber);
  return pluralTitle.substring(0, pluralTitle.length - 1);
};

export const otherChamber = (chamber) =>
  chamber === "house" ? "senate" : "house";
