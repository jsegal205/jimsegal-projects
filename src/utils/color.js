const getRGB = (hex) => {
  return {
    r: parseInt(hex.slice(0, 2), 16),
    g: parseInt(hex.slice(2, 4), 16),
    b: parseInt(hex.slice(4, 6), 16),
  };
};

export const randomHexColor = () =>
  Math.floor(Math.random() * 16777215).toString(16);

export const complementaryColor = (hex) => {
  const padZero = (str) => ("00" + str).slice(-2);
  const invertColor = (digit) => padZero((255 - digit).toString(16));
  const { r, g, b } = getRGB(hex);

  return invertColor(r) + invertColor(g) + invertColor(b);
};

export const textColor = (hex) => {
  const { r, g, b } = getRGB(hex);
  const brightness = Math.round((r * 299 + g * 587 + b * 114) / 1000);

  return brightness > 125 ? "444444" : "efefef";
};
