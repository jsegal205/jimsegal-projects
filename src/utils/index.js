// export const apiUrlBase = "https://api.jimsegal.com";
export const apiUrlBase = "http://localhost:8001";

export const projectInfo = [
  {
    absolute: true,
    link: "https://jimsegal.com",
    subtitle: "OG site with other projects",
    title: "JimSegal.com",
  },
  {
    link: "/congress",
    subtitle: "Some fun data surrounding the current congressional sesssion",
    title: "Congress",
  },
  {
    link: "/isJimWearingShorts",
    subtitle: "A site to predict if Jim is wearing shorts or not",
    title: "Is Jim Wearing Shorts?",
  },
  {
    link: "mountaingoat",
    subtitle: "Digital game",
    title: "Mountain Goat Game",
  },
  {
    link: "/recipes",
    subtitle: "A bunch of recipes that I frequent and wanted to show off",
    title: "Recipes",
  },
  {
    link: "/spacex",
    subtitle: "A countdown timer until the next launch of a Space X rocket",
    title: "Space X",
  },
];
