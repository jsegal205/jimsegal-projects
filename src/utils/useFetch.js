import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
