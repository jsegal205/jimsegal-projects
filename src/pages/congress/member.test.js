import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import { apiUrlBase } from "../../utils";
import useFetch from "../../utils/useFetch";
import Member from "./member";

jest.mock("../../utils/useFetch", () => jest.fn());

describe("Member Component", () => {
  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      const chamber = "house";
      const id = "1";
      const props = {
        location: {
          pathname: `/congress/${chamber}/member/${id}`,
        },
      };

      useFetch.mockReturnValue({
        loading: true,
        data: {},
      });

      const tree = renderer.create(
        <Router>
          <Member {...props} />
        </Router>
      );
      expect(useFetch).toHaveBeenCalledWith(
        `${apiUrlBase}/congress/${chamber}/member/${id}`
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    describe("with 404 status", () => {
      it("displays `Not Found`", () => {
        const chamber = "house";
        const id = "1";
        const props = {
          location: {
            pathname: `/congress/${chamber}/member/${id}`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: { status: 404 },
        });

        const tree = renderer.create(
          <Router>
            <Member {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/member/${id}`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });

    describe("with 500 status", () => {
      it("displays error message", () => {
        const chamber = "house";
        const id = "1";
        const props = {
          location: {
            pathname: `/congress/${chamber}/member/${id}`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: { error: "server likely on fire", status: 500 },
        });

        const tree = renderer.create(
          <Router>
            <Member {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/member/${id}`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });

    describe("with 200 status", () => {
      it("displays data", () => {
        const chamber = "house";
        const id = "1";
        const props = {
          location: {
            pathname: `/congress/${chamber}/member/${id}`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: {
            first_name: "first",
            last_name: "last",
            most_recent_vote: "most_recent_vote",
            age: 1,
            date_of_birth: "1900-01-01",
            gender: "F",
            current_party: "D",
            url: "www.example.com",
            twitter_account: "twitter",
            state: "IL",
            next_election: "2020",
            terms: 1,
            initial_elected_in: "2000-01-01",
            most_recent_vote: "2020-01-01",
            careerVoting: {
              careerVotesEligible: 1,
              careerVotesCast: 2,
              careerMissedVotes: 3,
              careerPresentVotes: 4,
              careerVotesWithParty: 5,
              careerVotesAgainstParty: 6,
            },
            termInfo: [
              {
                congress: "116",
                start_date: "2019-01-03",
                end_date: "2021-01-03",
                total_votes: 6,
                missed_votes: 2,
                total_present: 3,
                votesWithParty: 2,
                votesAgainstParty: 3,
              },
            ],
            status: 200,
          },
        });

        const tree = renderer.create(
          <Router>
            <Member {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/member/${id}`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe("getSlug", () => {
    describe("when props contain state.chamber and state.id", () => {
      it("pulls params from state", () => {
        const chamber = "house";
        const id = "1";
        const props = {
          location: {
            state: {
              chamber,
              id,
            },
            pathname: `/congress/somewhereelse/member/wrong`,
          },
        };

        useFetch.mockReturnValue({
          loading: true,
          data: {},
        });

        renderer.create(
          <Router>
            <Member {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/member/${id}`
        );
      });
    });
    describe("when props do not contain state.chamber or state.id", () => {
      it("pulls params from pathname", () => {
        const chamber = "house";
        const id = "1";
        const props = {
          location: {
            state: {
              "not-a-chamber": "state-chamber",
              "not-an-id": "state-id",
            },
            pathname: `/congress/${chamber}/member/${id}`,
          },
        };

        useFetch.mockReturnValue({
          loading: true,
          data: {},
        });

        renderer.create(
          <Router>
            <Member {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/member/${id}`
        );
      });
    });
  });
});
