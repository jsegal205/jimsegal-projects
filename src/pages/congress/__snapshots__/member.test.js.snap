// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<section
  className="loader"
>
  <div
    className="loader-wheel"
  />
  <div
    className="loader-text"
  >
    Loading
  </div>
</section>
`;

exports[`Member Component when \`useFetch\` returns data with 200 status displays data 1`] = `
<section
  className="congress-member"
>
  <a
    href="/congress/house/members"
    onClick={[Function]}
  >
    Back to all members
  </a>
  <h2>
    first
     
    last
  </h2>
  <div
    className="headshot"
  >
    <img
      alt="Professional headshot of first last"
      onError={[Function]}
      src="https://www.congress.gov/img/member/1.jpg"
    />
  </div>
  <section
    className="congress-member-links"
  >
    <a
      href="www.example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Website
    </a>
    <a
      href="https://twitter.com/twitter"
      rel="noopener noreferrer"
      target="_blank"
    >
      Twitter
    </a>
  </section>
  <hr />
  <article
    className="information"
  >
    <section>
      <ul>
        <li>
          <label>
            Most recent vote:
          </label>
           
          2020-01-01
        </li>
        <li>
          <label>
            Date of Birth:
          </label>
           
          1900-01-01
           (age: 
          1
          )
        </li>
        <li>
          <label>
            Gender:
          </label>
           
          F
        </li>
        <li>
          <label>
            Current Party:
          </label>
           
          Democrat
        </li>
        <li>
          <label>
            State Representing:
          </label>
           
          IL
        </li>
        <li>
          <label>
            Will be up for re-election in:
          </label>
           
          2020
        </li>
        <li>
          <p>
            Has served 
            <em>
              1
            </em>
             terms starting initially elected into office in 
            <em>
              2000-01-01
            </em>
          </p>
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Career Voting History
      </h3>
      <ul>
        <li>
          <label>
            Total Votes Eligible:
          </label>
           
          1
        </li>
        <li>
          <label>
            Total Cast Votes:
          </label>
           
          2
        </li>
        <li>
          <label>
            Total Missed Votes:
          </label>
           
          3
        </li>
        <li>
          <label>
            Total Present Votes:
          </label>
           
          4
        </li>
        <li>
          <label>
            Total Votes With 
            Democrat
             Party:
          </label>
           
          5
          *
        </li>
        <li>
          <label>
            Total Votes Against 
            Democrat
             Party:
          </label>
           
          6
          *
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Individual Term Voting Statistics
      </h3>
      <ul
        className="term-info"
      >
        <li>
          <label>
            Congressional Session:
          </label>
           
          116
           - From
           
          2019-01-03
           until 
          2021-01-03
        </li>
        <li>
          <label>
            Cast Votes:
          </label>
           
          1
        </li>
        <li>
          <label>
            Votes With 
            Democrat
             Party:
          </label>
           
          2
          *
        </li>
        <li>
          <label>
            Votes Against 
            Democrat
             Party:
          </label>
           
          3
          *
        </li>
      </ul>
    </section>
    <footer
      className="footnote"
    >
      * - counts relating to votes with or against party might be off due to rounding precision or insufficient historical data provided
    </footer>
  </article>
</section>
`;

exports[`Member Component when \`useFetch\` returns data with 200 status displays misconduct info if present 1`] = `
<section
  className="congress-member"
>
  <a
    href="/congress/house/members"
    onClick={[Function]}
  >
    Back to all members
  </a>
  <h2>
    first
     
    last
  </h2>
  <div
    className="headshot"
  >
    <img
      alt="Professional headshot of first last"
      onError={[Function]}
      src="https://www.congress.gov/img/member/1.jpg"
    />
  </div>
  <section
    className="congress-member-links"
  >
    <a
      href="www.example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Website
    </a>
    <a
      href="https://twitter.com/twitter"
      rel="noopener noreferrer"
      target="_blank"
    >
      Twitter
    </a>
  </section>
  <hr />
  <article
    className="information"
  >
    <section>
      <ul>
        <li>
          <label>
            Most recent vote:
          </label>
           
          2020-01-01
        </li>
        <li>
          <label>
            Date of Birth:
          </label>
           
          1900-01-01
           (age: 
          1
          )
        </li>
        <li>
          <label>
            Gender:
          </label>
           
          F
        </li>
        <li>
          <label>
            Current Party:
          </label>
           
          Democrat
        </li>
        <li>
          <label>
            State Representing:
          </label>
           
          IL
        </li>
        <li>
          <label>
            Will be up for re-election in:
          </label>
           
          2020
        </li>
        <li>
          <p>
            Has served 
            <em>
              1
            </em>
             terms starting initially elected into office in 
            <em>
              2000-01-01
            </em>
          </p>
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Official Misconduct Reports
      </h3>
      <h4>
        Data provided by
         
        <a
          href="https://github.com/govtrack/misconduct"
          rel="noopener noreferrer"
          target="_blank"
        >
          GovTrack Congressional Misconduct DB
        </a>
      </h4>
      <ul>
        <li>
          <label>
            Allegation:
          </label>
           
          Allegation text
        </li>
        <li>
          <label>
            Initiated on:
          </label>
           
          2020-01-01
        </li>
        <li>
          <label>
            Allegation Categories:
          </label>
           
          corruption, elections
        </li>
        <li>
          <label>
            Current Status:
          </label>
           
          Unresolved
        </li>
        <li>
          <p>
            text in 
            <a
              href="#markdown"
            >
              markdown
            </a>
          </p>
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Career Voting History
      </h3>
      <ul>
        <li>
          <label>
            Total Votes Eligible:
          </label>
           
          1
        </li>
        <li>
          <label>
            Total Cast Votes:
          </label>
           
          2
        </li>
        <li>
          <label>
            Total Missed Votes:
          </label>
           
          3
        </li>
        <li>
          <label>
            Total Present Votes:
          </label>
           
          4
        </li>
        <li>
          <label>
            Total Votes With 
            Democrat
             Party:
          </label>
           
          5
          *
        </li>
        <li>
          <label>
            Total Votes Against 
            Democrat
             Party:
          </label>
           
          6
          *
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Individual Term Voting Statistics
      </h3>
      <ul
        className="term-info"
      >
        <li>
          <label>
            Congressional Session:
          </label>
           
          116
           - From
           
          2019-01-03
           until 
          2021-01-03
        </li>
        <li>
          <label>
            Cast Votes:
          </label>
           
          1
        </li>
        <li>
          <label>
            Votes With 
            Democrat
             Party:
          </label>
           
          2
          *
        </li>
        <li>
          <label>
            Votes Against 
            Democrat
             Party:
          </label>
           
          3
          *
        </li>
      </ul>
    </section>
    <footer
      className="footnote"
    >
      * - counts relating to votes with or against party might be off due to rounding precision or insufficient historical data provided
    </footer>
  </article>
</section>
`;

exports[`Member Component when \`useFetch\` returns data with 200 status displays term party if different from current party 1`] = `
<section
  className="congress-member"
>
  <a
    href="/congress/house/members"
    onClick={[Function]}
  >
    Back to all members
  </a>
  <h2>
    first
     
    last
  </h2>
  <div
    className="headshot"
  >
    <img
      alt="Professional headshot of first last"
      onError={[Function]}
      src="https://www.congress.gov/img/member/1.jpg"
    />
  </div>
  <section
    className="congress-member-links"
  >
    <a
      href="www.example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      Website
    </a>
    <a
      href="https://twitter.com/twitter"
      rel="noopener noreferrer"
      target="_blank"
    >
      Twitter
    </a>
  </section>
  <hr />
  <article
    className="information"
  >
    <section>
      <ul>
        <li>
          <label>
            Most recent vote:
          </label>
           
          2020-01-01
        </li>
        <li>
          <label>
            Date of Birth:
          </label>
           
          1900-01-01
           (age: 
          1
          )
        </li>
        <li>
          <label>
            Gender:
          </label>
           
          F
        </li>
        <li>
          <label>
            Current Party:
          </label>
           
          Democrat
        </li>
        <li>
          <label>
            State Representing:
          </label>
           
          IL
        </li>
        <li>
          <label>
            Will be up for re-election in:
          </label>
           
          2020
        </li>
        <li>
          <p>
            Has served 
            <em>
              1
            </em>
             terms starting initially elected into office in 
            <em>
              2000-01-01
            </em>
          </p>
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Career Voting History
      </h3>
      <ul>
        <li>
          <label>
            Total Votes Eligible:
          </label>
           
          1
        </li>
        <li>
          <label>
            Total Cast Votes:
          </label>
           
          2
        </li>
        <li>
          <label>
            Total Missed Votes:
          </label>
           
          3
        </li>
        <li>
          <label>
            Total Present Votes:
          </label>
           
          4
        </li>
        <li>
          <label>
            Total Votes With 
            Democrat
             Party:
          </label>
           
          5
          *
        </li>
        <li>
          <label>
            Total Votes Against 
            Democrat
             Party:
          </label>
           
          6
          *
        </li>
      </ul>
    </section>
    <section>
      <h3>
        Individual Term Voting Statistics
      </h3>
      <ul
        className="term-info"
      >
        <li>
          <label>
            Congressional Session:
          </label>
           
          116
           - From
           
          2019-01-03
           until 
          2021-01-03
        </li>
        <li>
          <label>
            Cast Votes:
          </label>
           
          1
        </li>
        <li>
          <label>
            Votes With 
            Democrat
             Party:
          </label>
           
          2
          *
        </li>
        <li>
          <label>
            Votes Against 
            Democrat
             Party:
          </label>
           
          3
          *
        </li>
      </ul>
      <ul
        className="term-info"
      >
        <li>
          <label>
            Congressional Session:
          </label>
           
          115
           - From
           
          2017-01-03
           until 
          2019-01-03
        </li>
        <li>
          <label>
            Party:
          </label>
           
          Republican
        </li>
        <li>
          <label>
            Cast Votes:
          </label>
           
          1
        </li>
        <li>
          <label>
            Votes With 
            Republican
             Party:
          </label>
           
          2
          *
        </li>
        <li>
          <label>
            Votes Against 
            Republican
             Party:
          </label>
           
          3
          *
        </li>
      </ul>
    </section>
    <footer
      className="footnote"
    >
      * - counts relating to votes with or against party might be off due to rounding precision or insufficient historical data provided
    </footer>
  </article>
</section>
`;

exports[`Member Component when \`useFetch\` returns data with 404 status displays \`Not Found\` 1`] = `
<h1>
  Not Found
</h1>
`;

exports[`Member Component when \`useFetch\` returns data with 500 status displays error message 1`] = `
<section>
  <h3>
    Whoops! There was a problem loading
     
    Congress Memeber data.
  </h3>
  <p>
    Please reload browser to try again in a little bit.
  </p>
</section>
`;
